aws_lambda_powertools/utilities/data_classes/s3_batch_operation_event.py

```diff
@@ -0,0 +1,220 @@
+    @property
+    def task_id(self) -> str:
+        """Get the task id"""
+        return self["taskId"]
+
+    @property
+    def s3_key(self) -> str:
+        """Get the object key unquote_plus using strict utf-8 encoding"""
+        # note: AWS documentation example is using unquote but this actually
+        # contradicts what happens in practice. The key is url encoded with %20
+        # in the inventory file but in the event it is sent with +. So use unquote_plus
+        return unquote_plus(self["s3Key"], encoding="utf-8", errors="strict")
+
+    @property
+    def s3_version_id(self) -> Optional[str]:
+        """Object version if bucket is versioning-enabled, otherwise null"""
+        return self.get("s3VersionId")
+
+    @property
+    def s3_bucket_arn(self) -> Optional[str]:
+        """Get the s3 bucket arn (present only for invocationSchemaVersion '1.0')"""
+        return self.get("s3BucketArn")
+
+    @property
+    def s3_bucket(self) -> str:
+        """ "
+        Get the s3 bucket, either from 's3Bucket' property (invocationSchemaVersion '2.0')
+        or from 's3BucketArn' (invocationSchemaVersion '1.0')
+        """
+        if self.s3_bucket_arn:
+            return self.s3_bucket_arn.split(":::")[-1]
+        return self["s3Bucket"]
+
+
+class S3BatchOperationEvent(DictWrapper):
+    """Amazon S3BatchOperation Event
+
+    Documentation:
+    --------------
+    - https://docs.aws.amazon.com/AmazonS3/latest/userguide/batch-ops-invoke-lambda.html
+    """
+
+    @property
+    def invocation_id(self) -> str:
+        """Get the identifier of the invocation request"""
+        return self["invocationId"]
+
+    @property
+    def invocation_schema_version(self) -> str:
+        """ "
+        Get the schema version for the payload that Batch Operations sends when invoking an
+        AWS Lambda function. Either '1.0' or '2.0'.
+        """
+        return self["invocationSchemaVersion"]
+
+    @property
+    def tasks(self) -> Iterator[S3BatchOperationTask]:
+        for task in self["tasks"]:
+            yield S3BatchOperationTask(task)
+
+    @property
+    def task(self) -> S3BatchOperationTask:
+        """Get the first s3 batch operation task"""
+        return next(self.tasks)
+
+    @property
+    def job(self) -> S3BatchOperationJob:
+        """Get the s3 batch operation job"""
+        return S3BatchOperationJob(self["job"])
+
+
+# list of valid result code. Used both in S3BatchOperationResult and S3BatchOperationResponse
+VALID_RESULT_CODE_TYPES: Tuple[str, str, str] = ("Succeeded", "TemporaryFailure", "PermanentFailure")
+
+
+@dataclass(repr=False, order=False)
+class S3BatchOperationResult:
+    task_id: str
+    result_code: Literal["Succeeded", "TemporaryFailure", "PermanentFailure"]
+    result_string: Optional[str] = None
+
+    def __post_init__(self):
+        if self.result_code not in VALID_RESULT_CODE_TYPES:
+            raise ValueError(f"Invalid result_code: {self.result_code}")
```

Comment:
I think we shouldn't raise a Runtime exception if the customer uses an invalid result code. We should have a warning and the customer can fix that.

We have 2 results if the customer uses an invalid result code:

1 - If the customer removes the `treatMissingKeysAs` key, the S3 Batch will consider the execution successful and will consider the task completed successfully.
2 - If the client uses an invalid code such as `blablabla`, S3 Batch will consider that the execution failed and will add the reason to the report job.

As this is a batch execution, the customer may have a problem with business logic/idempotence/others if we create an exception. The customer should read the S3 Batch Report Job and correct possible problems.

What do you think?

Reply:
Late to the review, but it is up to you I would say to be consistent with the rest of the project. I was on the fence for this part of the code as I'm coming from more typesafe languages and also a big fan of early assertions that should happen ideally as close as possible to the original issue. I was thinking to use enums here but didn't went to write something not consistent with the kinesis classes that were already using literals, so I ended somewhat here not completely satisfied with the results. I'm not a fan in general of logs for things like this that tend to get too easy to miss, so that's the reason I preferred the exception.